core/src/main/scala/kafka/security/auth/SimpleAclAuthorizer.scala

```diff
@@ -104,26 +104,32 @@ class SimpleAclAuthorizer extends Authorizer with Logging {
   override def authorize(session: Session, operation: Operation, resource: Resource): Boolean = {
     val principal = session.principal
     val host = session.clientAddress.getHostAddress
-    val acls = getAcls(resource) ++ getAcls(new Resource(resource.resourceType, Resource.WildCardResource))
-
-    // Check if there is any Deny acl match that would disallow this operation.
-    val denyMatch = aclMatch(operation, resource, principal, host, Deny, acls)
-
-    // Check if there are any Allow ACLs which would allow this operation.
-    // Allowing read, write, delete, or alter implies allowing describe.
-    // See #{org.apache.kafka.common.acl.AclOperation} for more details about ACL inheritance.
-    val allowOps = operation match {
-      case Describe => Set[Operation](Describe, Read, Write, Delete, Alter)
-      case DescribeConfigs => Set[Operation](DescribeConfigs, AlterConfigs)
-      case _ => Set[Operation](operation)
-    }
-    val allowMatch = allowOps.exists(operation => aclMatch(operation, resource, principal, host, Allow, acls))
 
-    //we allow an operation if a user is a super user or if no acls are found and user has configured to allow all users
-    //when no acls are found or if no deny acls are found and at least one allow acls matches.
-    val authorized = isSuperUser(operation, resource, principal, host) ||
+    val authorized = if (isSuperUser(operation, resource, principal, host)) {
+      // Check if the user is a super user to avoid needlessly checking all ACLs
+      // for super users
+      true
+    } else {
+      val acls = getAcls(resource) ++ getAcls(new Resource(resource.resourceType, Resource.WildCardResource))
+
+      // Check if there is any Deny acl match that would disallow this operation.
+      val denyMatch = aclMatch(operation, resource, principal, host, Deny, acls)
+
+      // Check if there are any Allow ACLs which would allow this operation.
+      // Allowing read, write, delete, or alter implies allowing describe.
+      // See #{org.apache.kafka.common.acl.AclOperation} for more details about ACL inheritance.
+      val allowOps = operation match {
+        case Describe => Set[Operation](Describe, Read, Write, Delete, Alter)
+        case DescribeConfigs => Set[Operation](DescribeConfigs, AlterConfigs)
+        case _ => Set[Operation](operation)
+      }
+      val allowMatch = allowOps.exists(operation => aclMatch(operation, resource, principal, host, Allow, acls))
+
+      //we allow an operation if no acls are found and user has configured to allow all users
+      //when no acls are found or if no deny acls are found and at least one allow acls matches.
       isEmptyAclAndAuthorized(operation, resource, principal, host, acls) ||
```

Comment:
Would it make sense to move the empty check above the ACL matching since we don't really care about the results anyway in that case?

Reply:
Makes sense, I'll amend the PR.

Reply:
Actually after doing some more thinking and looking at the code it may make sense to leave the check as is.

The log message that is currently being displayed is:

    authorizerLogger.debug(s"No acl found for resource $resource, authorized = $shouldAllowEveryoneIfNoAclIsFound")

Which would not be factually correct any more if we made the switch, as we never looked at ACLs, so there may very well be ACLs as well that would permit the operation when this message is shown.

We could reword the log message, but I'd argue that the information that no ACLs matched can be useful when securing an already busy cluster to find clients that were missed in the migration (and probably other scenarios as well).

The overhead is also reasonable I think, we need to load the ACLs anyway, otherwise we can't perform the check and then its just two .exists checks against an empty collection that need to be performed (once for Deny matches once for Allow matches).

Reply:
Yeah I didn't see the harm since we would still check `isEmptyAclAndAuthorized` in any case. Not that big of a deal though.

Reply:
I've created a commit that addresses this comment as well.. I'm not sure that I like it much better than the code that is currently there though..we can't simply add the check as an "if else" block, since we need to retrieve the ACLs after the superuser check, so it becomes a nested if..
Also this created "dead" code in isEmptyAclAndAuthorized so I inlined that function as it was extremely short and I suspect it was only initially created to keep the conditional statement readable.

I've pushed to an [alternative branch](https://github.com/soenkeliebau/kafka/commit/6a2156144091142d948ce0c7b2e274744629936b) so I don't have to revert here. Feel free to have a look and let me know what you think.

Reply:
@hachikuji Did you have a chance to have a look at the proposed alternative and have any thoughts on whether we want to use that or keep the PR as is?

Reply:
@soenkeliebau The alternative looks good to me.